Hermite polynomials and transient flow solution

MahdiRoshani opened the pull request.

MahdiRoshani committed:
WIP: Hermite polynomials and transient flow solution

MahdiRoshani committed:
ENH: Tested transient solution for Euler–Bernoulli beam equation with constant load

MahdiRoshani committed:
TST: Added Pressure solution

MahdiRoshani committed:
TST: Beam solution for linear load

MahdiRoshani committed:
TST: Added flow rate solution

MahdiRoshani committed:
ENH: Two-way FSI coupling for one dimensional simulation

MahdiRoshani committed:
ENH: Updated p and H_new in Beam.py function to p_interleaved and H_new_interleaved

MahdiRoshani committed:
ENH: test passed for two-way FSI coupling for one dimensional simulation.

MahdiRoshani committed:
TST: Added residual plot for two-way FSI coupling for one dimensional simulation.

MahdiRoshani committed:
WIP: Two-way FSI coupling for one dimensional simulation is running on NEMO2

MahdiRoshani committed:
ENH: BasisFunctions.py and test_basis_functions.py enhanced

MahdiRoshani committed:
ENH: test_quadrature.py file enhanced

pastewka commented on VascularFlow/Coupling/OneDimensional.py:
```diff
@@ -0,0 +1,117 @@
+        outer_iteration_number += 1
+        inner_iteration_number = 0
+        inner_residual = 1
+        while inner_residual > 10e-08 and inner_iteration_number < 10000:
```
Please refactor the inner loop into its own function. The residual and number of iterations should become arguments of that function.

MahdiRoshani replied:
All done

pastewka commented on VascularFlow/Coupling/OneDimensional.py:
```diff
@@ -0,0 +1,117 @@
+            )
+
+            # height calculation
+            h_star = euler_bernoulli_transient(
```
It's confusing that this is called transient here. The FSI solution should be steady-state I suppose.

MahdiRoshani replied:
I used strong two-way coupling in the simulation. 
It is also possible to have a steady version of the FSI simulation. Which is studied by the paper I attached. 
[s00021-002-8536-9.pdf](https://github.com/user-attachments/files/19588507/s00021-002-8536-9.pdf)

pastewka commented on VascularFlow/Coupling/OneDimensional.py:
```diff
@@ -0,0 +1,117 @@
+            )
+
+            # update inner iteration
+            if max(abs(h_new - 1)) < 10e-20:
```
The tolerance should be an argument of the function, not hard-coded.

pastewka commented on VascularFlow/Elasticity/Beam.py:
```diff
@@ -22,42 +31,95 @@ def euler_bernoulli(x_n, q_g):
+    return system_matrix_gg, system_matrix_ll, w_g
+
+
+def euler_bernoulli_transient(x_n, dx_e, num_steps, dt, p, beta, relaxation, H_new):
```
This looks like this is doubling a lot of code that is already in the other Euler-Bernoulli function. Try to move that code to a separate utility function so it only appears once. Also please add a documentation string, I don't really understand what this function does. It is actually time-propagating the wave equation on the Euler-Bernoulli beam?

MahdiRoshani replied:
Added a function for the clamped boundary condition to avoid repeating lines

pastewka commented on VascularFlow/Initialization/InitializeConstants.py:
```diff
@@ -0,0 +1,15 @@
+def initialize_constants():
```
Do not put the simulation parameters in the module. They should go into a separate simulation script.

pastewka commented on VascularFlow/Initialization/InitializeFlowArrays.py:
```diff
@@ -0,0 +1,26 @@
+    """
+    size = nb_nodes
+    return {
+        "h_n_1": np.ones(size),
```
Do you actually need this many auxiliary fields? Can you put comments here explaining what each of these fields stores?

pastewka commented on VascularFlow/Numerics/ArrayFirstDerivative.py:
```diff
@@ -0,0 +1,16 @@
+import numpy as np
+
+def array_first_derivative(array: np.ndarray, dx_e: float) -> np.ndarray:
```
What does this function do? Please add a documentation string.

It looks like some finite differences scheme to me.

MahdiRoshani replied:
Added doc string

pastewka commented on VascularFlow/Numerics/Assembly.py:
```diff
@@ -18,3 +17,47 @@ def assemble_system_matrix(element_matrices_enn: ndarray):
     return system_matrix_gg
+
+
+def assemble_system_matrix_2dof(element_matrices_enn: ndarray):
```
Does this need to be a separate function? I'd prefer a single function with `dof`  as a parameter. I think it is just a question of changing a couple of indices.

MahdiRoshani replied:
Added one general function to create assembly matrix and vectors

pastewka commented on VascularFlow/Numerics/Assembly.py:
```diff
@@ -18,3 +17,47 @@ def assemble_system_matrix(element_matrices_enn: ndarray):
+    return system_matrix_gg
+
+
+def assemble_force_matrix_pressure(element_matrices_enn: ndarray):
```
Is this the force vector? (Then please rename function to assemble the vector.) How is this specific to pressure? Please also update the documentation string.

MahdiRoshani replied:
Added one general function to create assembly matrix and vectors, and removed all unnecessary functions

pastewka commented on VascularFlow/Numerics/ElementMatrices.py:
```diff
@@ -14,14 +15,48 @@ def outer_product(x_q: np.array) -> np.array:
 
     return gaussian_quadrature(nb_quad_pts, 0, 1, outer_product)
 
+def dx_matrix(dx: float):
```
Why is this necessary? The element matrices should be constructed from the basis functions.

MahdiRoshani replied:
Added one general function (load vector) to creta element vector

pastewka commented on VascularFlow/Numerics/ElementMatrices.py:
```diff
@@ -14,14 +15,48 @@ def outer_product(x_q: np.array) -> np.array:
-    )
+    ) * dx_matrix(dx)
+
+def force_matrix(dx: int):
```
Again here, please do not hard code matrices for a specific basis. The idea is to provide a basis set and use the generic quadrature method to compute the integrals.

MahdiRoshani replied:
Added one general function (load vector) to creta element vector

pastewka commented on VascularFlow/Numerics/ElementMatrices.py:
```diff
@@ -14,14 +15,48 @@ def outer_product(x_q: np.array) -> np.array:
+    return np.array([dx/2, dx**2/12, dx/2, -dx**2/12])
+
+
+def force_matrix_pressure(dx: int):
```
Same here.

MahdiRoshani replied:
Added one general function (load vector) to creta element vector

pastewka commented on tests/test_assembly.py:
```diff
@@ -25,13 +25,26 @@ def test_assemble_2x2():
+    system_matrix_gg = assemble_system_matrix_1dof(element_matrices_enn)
+    print(system_matrix_gg.shape)
+    print(system_matrix_gg)
+
```
There is some assertion missing here

MahdiRoshani replied:
Added assertion

pastewka commented on tests/test_beam.py:
```diff
@@ -1,22 +1,58 @@
+
+    w_g = euler_bernoulli(x_n, dx_e , p)[2]
+    exact = (x_n ** 4 / 24) - (x_n ** 3 / 12) + (x_n ** 2 / 24)
+    print(w_g.shape)
```
This is not yet a test. There is an assertion missing.

MahdiRoshani replied:
Added assertion

pastewka commented on tests/test_element_matrices.py:
```diff
@@ -1,22 +1,46 @@
+
+def test_force_matrix():
+    dx = 1
+    print(force_matrix(dx))
```
Not a test, assertion missing

MahdiRoshani replied:
Added one test (test_element_matrix_or_vector) for element matrices

pastewka commented on tests/test_element_matrices.py:
```diff
@@ -1,22 +1,46 @@
+
+def test_force_matrix_pressure():
+    dx = 1
+    print(force_matrix_pressure(dx))
```
Not a test, assertion missing

pastewka commented on tests/test_element_matrices.py:
```diff
@@ -1,22 +1,46 @@
+
+def test_dx_matrix_mass():
+    dx = 1
+    print(dx_matrix_mass(dx))
```
Not a test, assertion missing

pastewka commented on tests/test_element_matrices.py:
```diff
@@ -1,22 +1,46 @@
+    basis_function = HermiteBasis()
+    element_matrix = mass_matrix(3, 2, basis_function)
+    print(element_matrix.shape)
+    print(element_matrix)
```
Not a test, assertion missing

MahdiRoshani replied:
Added one test (test_element_matrix_or_vector) for element matrices

pastewka commented on tests/test_flow.py:
```diff
@@ -0,0 +1,89 @@
+    qq = flow_rate(
+        x_n, dx_e, dt, strouhal_number, inlet_flow_rate, channel_height, h_n, h_n_1
+    )
+    if plot:
```
Not a test, assertion missing

MahdiRoshani replied:
Added assertation

pastewka commented on tests/test_one_dimensional.py:
```diff
@@ -0,0 +1,74 @@
+    )
+    print(final_solution[0])
+    print(final_solution[1])
+    print(final_solution[2])
```
Not a test, assertion missing

MahdiRoshani replied:
Added a test for a range of physical and numerical parameters.

pastewka commented on tests/test_pressure.py:
```diff
@@ -0,0 +1,26 @@
+
+    pp = pressure(x_n, dx_e, dt, eps, re, st, Hstar, Qstar, Q_n, Q_n1)
+    print(pp.shape)
+    print(pp)
```
Not a test, assertion missing

MahdiRoshani replied:
Added the test (test_pressure)

MahdiRoshani committed:
ENH: ElementMatrices.py and test_element_matrices.py files enhanced

MahdiRoshani committed:
ENH: Added interpolation of first and second derivative of basis functions

MahdiRoshani committed:
TST: Test for the gaussian quadrature function for interpolated Hermite basis function

MahdiRoshani committed:
ENH: Updated the "ElementMatrices.py" to evaluate element matrices in an arbitrary interval. Also tested "test_element_matrices.py" for required element matrices

MahdiRoshani committed:
ENH: Added number of degree of freedom to BasisFunctions.py to update global matrix assembly for an arbitrary interval.

MahdiRoshani committed:
TST: Global matrix assembly for an arbitrary interval.

MahdiRoshani committed:
ENH: Added a function for clamped boundary condition used in Euler Bernoulli beam theory

MahdiRoshani committed:
TST: Clamped boundary condition used in Euler Bernoulli beam theory

MahdiRoshani committed:
ENH: Clarified Euler Bernoulli beam function

MahdiRoshani committed:
TST: Added assertation to test_beam.py

MahdiRoshani committed:
TST: Added a documentation string

MahdiRoshani committed:
ENH: Clarified the function for pressure

MahdiRoshani committed:
TST: Added assertation to test_pressure.py

MahdiRoshani committed:
ENH: Clarified the function for flow rate. And, made some changes to Pressure.py and test_pressure.py doc string

MahdiRoshani committed:
TST:  Added assertation to test_flow.py

MahdiRoshani committed:
ENH: Added inner loop for fsi

MahdiRoshani committed:
TST: Added a test for fsi inner loop

MahdiRoshani committed:
ENH: Clarified the function for two way fsi simulation

MahdiRoshani committed:
TST: Added assertation to test_one_dimensional.py

MahdiRoshani commented:
I am running  test_one_dimensional.py on NEMO2 again. For "inner_it_number=500," it takes too long to approach the next time step!!!

tests/test_one_dimensional.py Time: 0.00250, Inner Iteration: 500, Inner Residual: 3.88945e-04
Time: 0.00500, Inner Iteration: 500, Inner Residual: 1.71895e-04
Time: 0.00750, Inner Iteration: 500, Inner Residual: 1.59746e-04
Time: 0.01000, Inner Iteration: 500, Inner Residual: 4.56094e-04
Time: 0.01250, Inner Iteration: 500, Inner Residual: 1.00939e-03
Time: 0.01500, Inner Iteration: 500, Inner Residual: 1.47809e-03

MahdiRoshani committed:
ENH: Added some lines to the code to save the desired unknowns at each time step for making an animation

MahdiRoshani committed:
ENH:

MahdiRoshani committed:
ENH: Added a function to calculate steady state pressure in reduced order model

MahdiRoshani committed:
TST: Added a test for steady state pressure in reduced order model

MahdiRoshani committed:
ENH: Added a function for steady state Euler-Bernoulli beam equation solved by finite element method with Hermite basis function

MahdiRoshani committed:
ENH:

MahdiRoshani committed:
ENH: Added a test for steady state Euler-Bernoulli beam equation solved by finite element method with Hermite basis function

MahdiRoshani committed:
TST: Added a test for inner loop of unsteady fsi simulation for reduced order model

MahdiRoshani committed:
TST: Added assertation a for unsteady fsi simulation for reduced order model

MahdiRoshani committed:
ENH: FSI steady state simulation in reduced order model

MahdiRoshani committed:
TST: test for FSI steady state simulation in reduced order model

MahdiRoshani committed:
ENH: add 2D BilinearShapeFunctions class for Q1 quadrilateral elements

MahdiRoshani committed:
ENH: add 2D Gaussian quadrature rules for reference square. Implemented 2D quadrature points and weights for 4- and 9-point scheme

MahdiRoshani committed:
ENH: implement 2D element matrices and vectors for Q1 quadrilateral elements

MahdiRoshani committed:
ENH: add 2D mesh connectivity builder for Q1 quadrilateral elements

MahdiRoshani committed:
ENH: add 2D global assembly of FEM matrices and vectors

MahdiRoshani committed:
ENH: add 2D boundary DOF finder utility

- Implemented `boundary_dofs_2d` to return node indices for structured grids
- Supports boundaries: "bottom", "top", "left", "right"
- Handles both single and multiple boundary queries
- Returns results as tuples of DOF index lists for flexible use

MahdiRoshani committed:
ENH: implement 2D Dirichlet boundary condition application

MahdiRoshani committed:
ENH: add 2D Poisson equation solver (as an example) with Dirichlet BCs

MahdiRoshani committed:
TST: add tests for 2D Poisson equation solver with Q1 elements. Checked solver execution, global assembly. Added optional 2D plotting of solution field for visualization

MahdiRoshani committed:
TST: plot for 2D bilinear shape functions (Q1 elements)

MahdiRoshani committed:
TST: add test_quadrature.py for validating 2D Gaussian quadrature. Verified integration of sample polynomial functions with known analytical values

MahdiRoshani committed:
TST: add test_matrices.py for validating 2D element matrices. Added unit tests for mass and stiffness matrix assembly using bilinear Q1 shape function. Ensured integration with Gaussian quadrature produces stable 2D results

MahdiRoshani committed:
TST: add test_assembly.py for global 2D assembly routines

MahdiRoshani committed:
WIP: initial changes, will refine later

MahdiRoshani committed:
ENH: Adini-Clough-Melosh (ACM) shape functions for plate bending (Kirchhoff–Love plate equation)

MahdiRoshani committed:
ENH: Adini-Clough-Melosh (ACM) stiffness matrix and load vector for plate bending (Kirchhoff–Love plate equation)

MahdiRoshani committed:
ENH: Adini-Clough-Melosh (ACM) connectivity list (local-to-global node mapping matrix)   for plate bending (Kirchhoff–Love plate equation)

MahdiRoshani committed:
ENH: Adini-Clough-Melosh (ACM) boundary DOFs finder for applying boundary conditions for plate bending (Kirchhoff–Love plate equation)

MahdiRoshani committed:
ENH: Adini-Clough-Melosh (ACM) clamped boundary condition function for plate bending (Kirchhoff–Love plate equation)

MahdiRoshani committed:
ENH:  Assembly of the global stiffness matrix, and global source vector F for plate bending (Kirchhoff–Love plate equation)

MahdiRoshani committed:
ENH: plate bending (Kirchhoff–Love plate equation) solver

MahdiRoshani committed:
TST: plate bending (Kirchhoff–Love plate equation) solver

MahdiRoshani committed:
WIP: comparing the plate bending and beam bending solution

MahdiRoshani committed:
Merge branch 'main' into 25_hermite_and_flow

MahdiRoshani committed:
ENH: using "scysparse.lil_matrix" for the global stiffness matrix

MahdiRoshani committed:
Merge remote-tracking branch 'origin/25_hermite_and_flow' into 25_hermite_and_flow

MahdiRoshani committed:
ENH: changed the wrong transformation of second derivatives

MahdiRoshani committed:
ENH: changed the non-dimensional bending scale D and assembly for load vector
